ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1244 @@
 }
 
-int ossl_quic_clear(SSL *s)
+BIO *ossl_quic_conn_get_net_wbio(const QUIC_CONNECTION *qc)
+{
+    return qc->net_wbio;
+}
+
+int ossl_quic_conn_get_blocking_mode(const QUIC_CONNECTION *qc)
+{
+    return qc->blocking;
+}
+
+int ossl_quic_conn_set_blocking_mode(QUIC_CONNECTION *qc, int blocking)
 {
+    /* Cannot enable blocking mode if we do not have pollable FDs. */
+    if (blocking != 0 &&
+        (!qc->can_poll_net_rbio || !qc->can_poll_net_wbio))
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_UNSUPPORTED, NULL);
+
+    qc->blocking = (blocking != 0);
     return 1;
 }
 
-int ossl_quic_accept(SSL *s)
+int ossl_quic_conn_set_initial_peer_addr(QUIC_CONNECTION *qc,
+                                         const BIO_ADDR *peer_addr)
 {
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    if (qc->started)
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED,
+                                           NULL);
 
-    if (sc == NULL)
-        return 0;
+    if (peer_addr == NULL) {
+        BIO_ADDR_clear(&qc->init_peer_addr);
+        return 1;
+    }
 
-    ossl_statem_set_in_init(sc, 0);
+    qc->init_peer_addr = *peer_addr;
     return 1;
 }
 
-int ossl_quic_connect(SSL *s)
+/*
+ * QUIC Front-End I/O API: Asynchronous I/O Management
+ * ===================================================
+ *
+ *   (BIO/)SSL_tick                 => ossl_quic_tick
+ *   (BIO/)SSL_get_tick_timeout     => ossl_quic_get_tick_timeout
+ *   (BIO/)SSL_get_poll_fd          => ossl_quic_get_poll_fd
+ *
+ */
+
+/* Returns 1 if the connection is being used in blocking mode. */
+static int blocking_mode(const QUIC_CONNECTION *qc)
 {
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    return qc->blocking;
+}
 
-    if (sc == NULL)
-        return 0;
+/* SSL_tick; ticks the reactor. */
+int ossl_quic_tick(QUIC_CONNECTION *qc)
+{
+    if (qc->ch == NULL)
+        return 1;
 
-    ossl_statem_set_in_init(sc, 0);
+    ossl_quic_reactor_tick(ossl_quic_channel_get_reactor(qc->ch));
     return 1;
 }
 
-int ossl_quic_read(SSL *s, void *buf, size_t len, size_t *readbytes)
+/*
+ * SSL_get_tick_timeout. Get the time in milliseconds until the SSL object
+ * should be ticked by the application by calling SSL_tick(). tv is set to 0 if
+ * the object should be ticked immediately and tv->tv_sec is set to -1 if no
+ * timeout is currently active.
+ */
+int ossl_quic_get_tick_timeout(QUIC_CONNECTION *qc, struct timeval *tv)
 {
-    int ret;
-    BIO *rbio = SSL_get_rbio(s);
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    OSSL_TIME now, deadline = ossl_time_infinite();
+
+    if (qc->ch != NULL)
+        deadline
+            = ossl_quic_reactor_get_tick_deadline(ossl_quic_channel_get_reactor(qc->ch));
+
+    if (ossl_time_is_infinite(deadline)) {
+        tv->tv_sec  = -1;
+        tv->tv_usec = 0;
+        return 1;
+    }
+
+    now = ossl_time_now();
+    if (ossl_time_compare(now, deadline) >= 0) {
```

Comment:
Is this compare actually necessary? Will `ossl_time_substract` do the right thing anyway?